File: verify_receipt.py
#!/usr/bin/env python3
"""Verify a VerityLayer signed receipt OFFLINE. Imports nothing from VerityLayer.

    python verify_receipt.py receipt.json pubkey.json
    python verify_receipt.py receipt.json pubkey.json --claim "the claim it should vouch for"

Needs only the standard library plus `cryptography` (pip install cryptography).

The point of this file is that you should not have to trust us to check our own work. It
makes no network call, and it deliberately shares no code with the service that issues the
receipts - if the two agreed because they were the same code, that would prove nothing.

Get the two inputs once:
    curl -o receipt.json https://api.veritylayer.dev/receipt/selftest
    curl -o pubkey.json  https://api.veritylayer.dev/.well-known/verity-pubkey.json

Full spec: RECEIPTS.md in this repo.
Exit code 0 = signature valid (and bound, if --claim was given). Non-zero otherwise.
"""
from __future__ import annotations

import hashlib
import json
import sys


def canonical_bytes(receipt: dict) -> bytes:
    """The exact bytes that were signed.

    Sorted keys, no whitespace, `signature` excluded, UTF-8, non-ASCII left literal.

    NOTE: integral numbers serialize WITHOUT a decimal point - `confidence: 1` is `1`, not
    `1.0`. A language that renders every number as a float will produce different bytes and
    fail on a perfectly valid receipt. This is a deliberate divergence from RFC 8785 and is
    documented rather than left as a trap.
    """
    body = {k: v for k, v in receipt.items() if k != "signature"}
    return json.dumps(body, sort_keys=True, separators=(",", ":"),
                      ensure_ascii=False).encode("utf-8")


def derive_key_id(public_key_hex: str) -> str:
    """key_id is derived from the key, so you can confirm you hold the right one yourself."""
    return "ed25519:" + hashlib.sha256(bytes.fromhex(public_key_hex)).hexdigest()[:16]


def verify(receipt: dict, public_key_hex: str) -> tuple[bool, str]:
    try:
        from cryptography.exceptions import InvalidSignature
        from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
    except ImportError:
        return False, "this verifier needs `cryptography` - pip install cryptography"

    for field in ("signature", "key_id"):
        if not isinstance(receipt.get(field), str):
            return False, f"not a receipt: missing {field}"

    derived = derive_key_id(public_key_hex)
    if receipt["key_id"] != derived:
        # Do NOT 'helpfully' fetch a different key here. Falling back to asking the issuer
        # for a key that works is exactly the trust this file exists to avoid.
        return False, (f"wrong key: receipt was signed by {receipt['key_id']}, the key you "
                       f"supplied is {derived}")
    try:
        pub = Ed25519PublicKey.from_public_bytes(bytes.fromhex(public_key_hex))
        pub.verify(bytes.fromhex(receipt["signature"]), canonical_bytes(receipt))
    except InvalidSignature:
        return False, "SIGNATURE INVALID - this content is not what was signed"
    except Exception as e:  # malformed hex, wrong key length, etc.
        return False, f"could not verify: {type(e).__name__}: {e}"
    return True, "signature valid"


def check_binding(receipt: dict, claim: str) -> tuple[bool, str]:
    """A valid signature says WE SIGNED THIS. It does not say this vouches for YOUR claim."""
    want = hashlib.sha256(claim.encode("utf-8")).hexdigest()
    got = receipt.get("claim_sha256")
    if want == got:
        return True, "this receipt was issued for the claim you supplied"
    return False, ("SIGNATURE VALID BUT FOR A DIFFERENT CLAIM - it does not vouch for the "
                   "claim you supplied")


def _load(path: str) -> dict:
    with open(path, encoding="utf-8") as f:
        d = json.load(f)
    # /receipt/selftest wraps the receipt; accept either shape.
    return d.get("sample_receipt") or d.get("receipt") or d


def main(argv: list[str]) -> int:
    if len(argv) < 2:
        print(__doc__)
        return 2
    receipt = _load(argv[0])
    with open(argv[1], encoding="utf-8") as f:
        pub_hex = json.load(f)["public_key_hex"]

    ok, why = verify(receipt, pub_hex)
    print(f"signature : {'VALID' if ok else 'INVALID'} - {why}")
    if not ok:
        return 1

    if receipt.get("test") is True:
        print("warning   : this is a SELF-TEST receipt, not a paid verdict - it proves the "
              "signing chain works and nothing about any real claim")

    print(f"verdict   : {receipt.get('verdict')}  (confidence {receipt.get('confidence')})")
    print(f"endpoint  : {receipt.get('endpoint')}   issued {receipt.get('issued_at')}")

    if "--claim" in argv:
        claim = argv[argv.index("--claim") + 1]
        bound, why_b = check_binding(receipt, claim)
        print(f"binding   : {'BOUND' if bound else 'NOT BOUND'} - {why_b}")
        if not bound:
            return 1
    else:
        print("binding   : not checked - pass --claim \"...\" to confirm this receipt "
              "vouches for the claim you actually care about")
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))

File: verify_receipt.js
#!/usr/bin/env node
/**
 * Verify a VerityLayer signed receipt OFFLINE. Zero dependencies - Node's built-in crypto only.
 *
 *   node verify_receipt.js receipt.json pubkey.json
 *   node verify_receipt.js receipt.json pubkey.json --claim "the claim it should vouch for"
 *
 * Makes no network call and shares no code with the service that issues the receipts. If the
 * two agreed because they were the same code, that would prove nothing.
 *
 *   curl -o receipt.json https://api.veritylayer.dev/receipt/selftest
 *   curl -o pubkey.json  https://api.veritylayer.dev/.well-known/verity-pubkey.json
 *
 * Full spec: RECEIPTS.md in this repo.
 * Exit 0 = signature valid (and bound, if --claim given). Non-zero otherwise.
 */
const fs = require("fs");
const crypto = require("crypto");

/**
 * The exact bytes that were signed: sorted keys, no whitespace, `signature` excluded, UTF-8.
 *
 * JSON.stringify already renders integral numbers without a decimal point (1 -> "1", not
 * "1.0"), which is what the signer does - so this matches. It also emits non-ASCII literally,
 * matching ensure_ascii=false. Key order below is by code point via Array.sort()'s default
 * comparator on the code-unit level; identical for every field name in use.
 */
function canonicalBytes(receipt) {
  const body = {};
  for (const k of Object.keys(receipt).filter((k) => k !== "signature").sort()) body[k] = receipt[k];
  return Buffer.from(JSON.stringify(body), "utf8");
}

/** key_id is derived from the key, so you can confirm you hold the right one yourself. */
function deriveKeyId(publicKeyHex) {
  return "ed25519:" + crypto.createHash("sha256").update(Buffer.from(publicKeyHex, "hex")).digest("hex").slice(0, 16);
}

function verify(receipt, publicKeyHex) {
  for (const f of ["signature", "key_id"]) {
    if (typeof receipt[f] !== "string") return [false, `not a receipt: missing ${f}`];
  }
  const derived = deriveKeyId(publicKeyHex);
  if (receipt.key_id !== derived) {
    // Do NOT 'helpfully' fetch a key that works. That is the trust this file avoids.
    return [false, `wrong key: receipt signed by ${receipt.key_id}, you supplied ${derived}`];
  }
  try {
    // Wrap the raw 32-byte key as SPKI DER so crypto can read it; the prefix is the fixed
    // Ed25519 AlgorithmIdentifier.
    const der = Buffer.concat([Buffer.from("302a300506032b6570032100", "hex"),
                               Buffer.from(publicKeyHex, "hex")]);
    const pub = crypto.createPublicKey({ key: der, format: "der", type: "spki" });
    const ok = crypto.verify(null, canonicalBytes(receipt), pub, Buffer.from(receipt.signature, "hex"));
    return ok ? [true, "signature valid"]
              : [false, "SIGNATURE INVALID - this content is not what was signed"];
  } catch (e) {
    return [false, `could not verify: ${e.message}`];
  }
}

/** A valid signature says WE SIGNED THIS. Not that it vouches for YOUR claim. */
function checkBinding(receipt, claim) {
  const want = crypto.createHash("sha256").update(Buffer.from(claim, "utf8")).digest("hex");
  return want === receipt.claim_sha256
    ? [true, "this receipt was issued for the claim you supplied"]
    : [false, "SIGNATURE VALID BUT FOR A DIFFERENT CLAIM - it does not vouch for the claim you supplied"];
}

function load(p) {
  const d = JSON.parse(fs.readFileSync(p, "utf8"));
  return d.sample_receipt || d.receipt || d;   // /receipt/selftest wraps it
}

function main(argv) {
  if (argv.length < 2) { console.log(require("fs").readFileSync(__filename, "utf8").split("*/")[0]); return 2; }
  const receipt = load(argv[0]);
  const pubHex = JSON.parse(fs.readFileSync(argv[1], "utf8")).public_key_hex;

  const [ok, why] = verify(receipt, pubHex);
  console.log(`signature : ${ok ? "VALID" : "INVALID"} - ${why}`);
  if (!ok) return 1;

  if (receipt.test === true) {
    console.log("warning   : this is a SELF-TEST receipt, not a paid verdict - it proves the " +
                "signing chain works and nothing about any real claim");
  }
  console.log(`verdict   : ${receipt.verdict}  (confidence ${receipt.confidence})`);
  console.log(`endpoint  : ${receipt.endpoint}   issued ${receipt.issued_at}`);

  const i = argv.indexOf("--claim");
  if (i !== -1) {
    const [bound, whyB] = checkBinding(receipt, argv[i + 1]);
    console.log(`binding   : ${bound ? "BOUND" : "NOT BOUND"} - ${whyB}`);
    if (!bound) return 1;
  } else {
    console.log('binding   : not checked - pass --claim "..." to confirm this receipt ' +
                "vouches for the claim you actually care about");
  }
  return 0;
}

process.exit(main(process.argv.slice(2)));
